src/common/snippets/src/lowered/loop_manager.cpp

```diff
@@ -207,7 +214,169 @@ void LinearIR::LoopManager::mark_loop(LinearIR::constExprIt loop_begin_pos,
+                       LoopPort copy = loop_port;  // to save loop port parameters
+                       copy.expr_port = std::make_shared<ExpressionPort>(expr_port);
+                       return copy;
+                   });
+    port_it = ports.erase(port_it);
+    ports.insert(port_it, target_ports.cbegin(), target_ports.cend());
+}
 
+void LinearIR::LoopManager::update_loop_port(size_t loop_id, const LoopPort& actual_port, const std::vector<LoopPort>& target_ports,
+                                             bool is_entry) {
+    const auto& loop_info = get_loop_info(loop_id);
+    auto& ports = is_entry ? loop_info->entry_points : loop_info->exit_points;
+    auto port_it = std::find_if(ports.begin(), ports.end(),
+                                [&actual_port](const LoopPort& point) { return point == actual_port; });
+    OPENVINO_ASSERT(port_it != ports.end(), "Failed update_loop_port: existing loop ports has not been found");
+    port_it = ports.erase(port_it);
+    ports.insert(port_it, target_ports.cbegin(), target_ports.cend());
+}
+
+void LinearIR::LoopManager::expression_replacement(constExprIt new_expr_begin, constExprIt new_expr_end, const ExpressionPtr& decomposed_expr,
+                                                   size_t loop_id, const std::vector<ExpressionPort>& entries, const std::vector<ExpressionPort>& exits) {
+    for (auto it = new_expr_begin; it!= new_expr_end; ++it) {
+        insert_loop_id(*it, loop_id, true);
+    }
+    remove_loop_id(decomposed_expr, loop_id);
+
+    auto new_entries = entries;
+    auto new_exits = exits;
+    if (new_entries.empty() || new_exits.empty()) {
+        const auto loop_info = get_loop_info(loop_id);
+        get_io_loop_ports(new_expr_begin, new_expr_end, new_entries, new_exits);
+    }
+    for (size_t i = 0; i < decomposed_expr->get_input_count(); ++i) {
+        update_loop_port(loop_id, decomposed_expr->get_input_port(i), new_entries);
+    }
+    for (size_t i = 0; i < decomposed_expr->get_output_count(); ++i) {
+        update_loop_port(loop_id, decomposed_expr->get_output_port(i), new_exits, false);
+    }
+}
+
+void LinearIR::LoopManager::sort_loop_ports(LinearIR::constExprIt& loop_begin_pos, LinearIR::constExprIt& loop_end_pos, size_t loop_id) {
+    auto push = [](const std::vector<LoopPort>& ports, std::vector<LoopPort>& sorted_ports, const ExpressionPtr& expr) {
+        for (const auto& port : ports) {
+            if (port.expr_port->get_expr() == expr) {
+                sorted_ports.push_back(port);
+            }
+        }
+    };
+    auto loop_info = get_loop_info(loop_id);
+    const auto& loop_entries = loop_info->entry_points;
+    const auto& loop_exits = loop_info->exit_points;
+    std::vector<LoopPort> entries, exits;
+    entries.reserve(loop_entries.size());
+    exits.reserve(loop_exits.size());
+    for (auto it = loop_begin_pos; it != loop_end_pos; ++it) {
+        const auto& expr = *it;
+        push(loop_entries, entries, expr);
```

Comment:
Minor: It looks like this sort complexity is O(n2) in the worst case, where n is the number of expressions in the loop. This could be relaxed to O(n*log(n)) if we had float expression ids.